### PR TITLE
feat: add OpenSearch #21

### PR DESCRIPTION
**Pull request by cmltaWt0**
OpenSearch integration as a Shared resource.
===

Testing instructions
---
1. Install harmony chart.
2. Install and enable k8s_harmony tutor plugin.
3. Create opensearch user.
```
tutor harmony create-opensearch-user
```

4. Verify created User from an opensearch node:
```
curl --insecure -u harmony:${HARMONY_PASSWORD} -XGET https://localhost:9200/_plugins/_security/api/internalusers/openedx-01
```
result:
```
{"openedx-01":{"hash":"","reserved":false,"hidden":false,"backend_roles":[],"attributes":{},"opendistro_security_roles":["openedx-01_role"],"static":false}}
```
5. Verify created Role from an OpenSearch node:
```
 curl --insecure -u harmony:${HARMONY_PASSWORD} -XGET https://localhost:9200/_plugins/_security/api/roles/openedx-01_role
```
result:
```
{"openedx-01_role":{"reserved":false,"hidden":false,"cluster_permissions":[],"index_permissions":[{"index_patterns":["openedx-01-uwur-*"],"fls":[],"masked_fields":[],"allowed_actions":["all"]}],"tenant_permissions":[],"static":false}}
```
6. Install openedx via tutor k8s.
8. Open CMS pod.
9. Reindex content library
```
./manage.py cms reindex_content_library
```
11. Open Studio > Create a Course > Reindex it.

Implements #21 

Issues
---
- edx-platform has several places where Elastic is accessed directly w/o `edx-search` library. This leads to AccessDenied response due to Index prefix separation.
ex.
```
File "/openedx/edx-platform/cms/djangoapps/contentstore/management/commands/reindex_course.py", line 81, in handle
    index_exists = searcher._es.indices.exists(index=index_name)  # pylint: disable=protected-access
    ....
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, '')
```

### CODE REVIEW

**Comment by bradenmacdonald** [comment]
Awesome, thanks @cmltaWt0! let us know when you want a review.

**Comment by bradenmacdonald** [comment]
@cmltaWt0 Awesome, thanks! I probably can't get to this until next week but I'll take a look then.

> how to deal with enabled shared Elastic and OpenSearch at the same time.

I don't think we need to support that? People should be using one or the other.

**Review by MoisesGSalas — COMMENTED** [review]
I agree with Braden's comment about not supporting both opensearch and elasticsearch at the same time.

Would it be possible to use a single endpoint for the cluster using the
`clusterName` value? Seems like it's available in both the [opensearch](https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch/README.md?plain=1#L42) and [elasticsearch](https://github.com/elastic/helm-charts/blob/main/elasticsearch/README.md?plain=1#L130) charts. Setting that to a common value in our values.yaml for both `opensearch.clusterName` and `elasticsearch.clusterName` should do the trick I would think. Or is there any concern if we do that?

On a similar note, can we use a single helper class to interact with both APIs? The code in both `elasticsearch.py` and `opensearch.py` seems quite similar, so maybe instead we can use a single class that receives additional parameters besides `namespace` (like the endpoint for example) and each command instantiates it accordingly.

**Comment by cmltaWt0** [comment]
> 

I agree. Looks like the current simple implementation is enough.

**Comment by cmltaWt0** [comment]
> I agree with Braden's comment about not supporting both opensearch and elasticsearch at the same time.
> 
> Would it be possible to use a single endpoint for the cluster using the `clusterName` value? Seems like it's available in both the [opensearch](https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch/README.md?plain=1#L42) and [elasticsearch](https://github.com/elastic/helm-charts/blob/main/elasticsearch/README.md?plain=1#L130) charts. Setting that to a common value in our values.yaml for both `opensearch.clusterName` and `elasticsearch.clusterName` should do the trick I would think. Or is there any concern if we do that?
> 
> On a similar note, can we use a single helper class to interact with both APIs? The code in both `elasticsearch.py` and `opensearch.py` seems quite similar, so maybe instead we can use a single class that receives additional parameters besides `namespace` (like the endpoint for example) and each command instantiates it accordingly.

@MoisesGSalas Good idea, btw. 
I'll unite the entrypoint (cluster name) and try to check for incidentally enabling both elasticsearch and opensearch - what we want to prevent.

**Comment by cmltaWt0** [comment]
> The code in both elasticsearch.py and opensearch.py seems quite similar, so maybe instead we can use a single class that receives additional parameters besides namespace (like the endpoint for example) and each command instantiates it accordingly.

Definitely! Thanks.

**Comment by cmltaWt0** [comment]
@MoisesGSalas @bradenmacdonald I've added some sort of unification. Could we continue the review?

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings
Comment (MoisesGSalas) [inline tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings:5]: The service name that I got was actually `harmony-search-cluster.{{K8S_HARMONY_NAMESPACE}}.svc.cluster.local` instead of `harmony-search-cluster-master.{{K8S_HARMONY_NAMESPACE}}.svc.cluster.local`. After changing this I was able to index a course successfully.

**Review by MoisesGSalas — COMMENTED** [review]
Hi @cmltaWt0, while trying to deploy this in an EKS cluster I noticed a few errors at the start:

```
vm.max_map_count [65530] is too low, increase to at least [262144]
```

Seems like configuring that on opensearch is [disabled by default](https://github.com/opensearch-project/helm-charts/blob/main/charts/opensearch/values.yaml#L458-L463) while in the elasticsearch is [enabled by default](https://github.com/elastic/helm-charts/blob/main/elasticsearch/values.yaml#L295-L296). Maybe we should enabled it in our chart defaults? I didn't manage to make it work in my first test so I copy pasted [this](https://github.com/opensearch-project/helm-charts/issues/87#issuecomment-939127256), but it was probably an mistake on my part.

**Comment by itsjeyd** [comment]
Hey @cmltaWt0, a friendly reminder to follow up on @MoisesGSalas's comments above.

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings
Comment (cmltaWt0) [reply tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings:5]: Thanks @MoisesGSalas. I tested with these config successfully. Will re-test.

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py
Comment (cmltaWt0) [reply tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py:24]: Good catch! Thanks.

**Comment by cmltaWt0** [comment]
> this

@MoisesGSalas 
Will try to make it work and compare different solutions. Thanks.

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py
Comment (cmltaWt0) [reply tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/plugin.py]: Fixed

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings
Comment (cmltaWt0) [reply tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-common-settings]: Missed change after unification.
Fixed.

**Comment by cmltaWt0** [comment]
@MoisesGSalas 

> vm.max_map_count [65530] is too low, increase to at least [262144]

Also can't get it work following the documentation.
Working on it.

**Comment by cmltaWt0** [comment]
Found an issue with `reindex_course` command (some requests passed directly w/o a PREFIX):

```
./manage.py cms reindex_course --all --setup

...

2023-06-11 20:59:24,194 INFO 39 [elasticsearch] [user None] [ip None] base.py:270 - HEAD https://harmony-search-cluster.harmony.svc.cluster.local:9200/openedx-01-yts1-courseware_content [status:200 request:0.043s]
2023-06-11 20:59:24,201 WARNING 39 [elasticsearch] [user None] [ip None] base.py:293 - HEAD https://harmony-search-cluster.harmony.svc.cluster.local:9200/courseware_content [status:403 request:0.006s]
Traceback (most recent call last):
  File "./manage.py", line 106, in <module>
    execute_from_command_line([sys.argv[0]] + django_args)
  File "/openedx/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/openedx/venv/lib/python3.8/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/openedx/venv/lib/python3.8/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/openedx/venv/lib/python3.8/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/openedx/edx-platform/cms/djangoapps/contentstore/management/commands/reindex_course.py", line 81, in handle
    index_exists = searcher._es.indices.exists(index=index_name)  # pylint: disable=protected-access
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/client/utils.py", line 168, in _wrapped
    return func(*args, params=params, headers=headers, **kwargs)
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/client/indices.py", line 332, in exists
    return self.transport.perform_request(
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 415, in perform_request
    raise e
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/transport.py", line 381, in perform_request
    status, headers_response, data = connection.perform_request(
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/connection/http_urllib3.py", line 277, in perform_request
    self._raise_error(response.status, raw_data)
  File "/openedx/venv/lib/python3.8/site-packages/elasticsearch/connection/base.py", line 330, in _raise_error
    raise HTTP_EXCEPTIONS.get(status_code, TransportError)(
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, '')
```

Quick search revealed the following direct access of elastic indices from the management command [here](https://github.com/openedx/edx-platform/blob/dea67f29c41e588c297973a96c522af9a25d653d/cms/djangoapps/contentstore/management/commands/reindex_course.py#L81):
```
index_exists = searcher._es.indices.exists(index=index_name)
```

Need to test the fix.

**Comment by cmltaWt0** [comment]
> 

@MoisesGSalas 

I've managed to make it work with just enabling `sysctlInit`:
```
sysctlInit:
    enabled: true
```

That's add a new `sysctl` init container which doing the following:
```
+ DESIRED=262144
+ sysctl -n vm.max_map_count
+ CURRENT=262144
+ '[' 262144 -gt 262144 ]
```

Tested on minicube so need to re-test on EKS.

**Comment by cmltaWt0** [comment]
@MoisesGSalas it's a bit hardcoded [here](https://github.com/openedx/openedx-k8s-harmony/pull/31/files#diff-f9619be8f66226bf04af4e18a3d4ef0d7f4c8c8567ea21b4f16fbcd42c8b176dR3) to be able to test everything else. And I'm come up with a final configuration for certificates.

**Comment by itsjeyd** [comment]
@cmltaWt0 Thanks for the updates! Are the changes ready for another round of review or will you be making further changes to the code?

CC @MoisesGSalas

**Comment by cmltaWt0** [comment]
@itsjeyd PR is ready for the second review.
I've committed all changes I want. Thanks for ping!

**Comment by itsjeyd** [comment]
No problem @cmltaWt0.

@MoisesGSalas Could you please give this another look (or let us know when you think you'll have time for a second review pass)?

**Comment by MoisesGSalas** [comment]
@cmltaWt0, I think the issue with the reindex shouldn't be a blocker considering the --setup flag seems to be only used while using the devstack. I think the rest of the pr looks fine.

**Comment by MoisesGSalas** [comment]
@cmltaWt0, actually I've noticed an error while recreating the opensearch cluster and testing again. Whenever I try to refresh a course index I get a 403 error on the cms logs and the following error on the opensearch logs:

```log
[2023-06-27T22:11:22,024][INFO ][o.o.s.p.PrivilegesEvaluator] [harmony-search-cluster-master-1] No index-level perm match for User [name=openedx-01, backend_roles=[], requestedTenant=null] Resolved [aliases=[], allIndices=[courseware_content], types=[*], originalRequested=[courseware_content], remoteIndices=[]] [Action [indices:admin/get]] [RolesChecked [own_index, openedx-01_role]]
[2023-06-27T22:11:22,024][INFO ][o.o.s.p.PrivilegesEvaluator] [harmony-search-cluster-master-1] No permissions for [indices:admin/get]
```

I don't know if I missed something.

**Comment by cmltaWt0** [comment]
> ./manage.py cms reindex_course --all --setup

You didn't miss anything. That's exactly the same issue I described [here](https://github.com/openedx/openedx-k8s-harmony/pull/31#issuecomment-1586339576).

Your logs says the `openedx-01` is trying to access the `originalRequested=[courseware_content]` index without any prefix.
So you probably encountered another place in edx-platform the elasticsearch is accessed without edx-search and asking for index without a prefix.

Sample log with a prefixed index:
```
[2023-06-28T07:49:21,985][INFO ][o.o.p.PluginsService     ] [harmony-search-cluster-master-0] PluginService:onIndexModule index:[openedx-01-yts1-courseware_content/dLsvVeyATfWRIM1By-OibA]
```

Could you show exactly how you do the reindexing?

**Comment by cmltaWt0** [comment]
@MoisesGSalas 

Could you please check whether you image uses open-craft original [version](https://github.com/open-craft/edx-search/tree/keith/prefixed-index-names) or `edx-search==3.5.0`?
They use different settings for some reason so please ensure you are using edx-search==3.5.0 and you django setting called   `ELASTIC_SEARCH_INDEX_PREFIX `.

I changed the setting name [here](https://github.com/openedx/openedx-k8s-harmony/pull/31/files#diff-ac390fccc9c6b23dff5a18f5b07300d19cb87ecc349cae4c0b52a302791c825aR1):
```
- ELASTICSEARCH_INDEX_PREFIX = "{{ELASTICSEARCH_INDEX_PREFIX}}"
+ ELASTIC_SEARCH_INDEX_PREFIX = "{{HARMONY_SEARCH_INDEX_PREFIX}}"
```

This is due the original work (open-craft/edx-search.git@keith/prefixed-index-names) uses [ELASTICSEARCH_INDEX_PREFIX](https://github.com/open-craft/edx-search/commit/8c618005d507fb504565cb55c7c64ebc3fbcb358#diff-c06ca8d1964c523eebc9c45b3706891cf57efaced8532e66155215e92ac47173R339) so our early work with shared [elasticsearch](https://github.com/openedx/openedx-k8s-harmony/pull/13/files#diff-409f2f23c942478fca990f907b6916e015564c4861d5be1dd2269fd7fb0e0c90R2) was using the same naming.

File: tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-dockerfile-post-python-requirements
Comment (cmltaWt0) [reply tutor-contrib-harmony-plugin/tutor_k8s_harmony_plugin/patches/openedx-dockerfile-post-python-requirements]: @MoisesGSalas Updated and re-tested with Palm with any patching. Everything works pretty good.

**Comment by cmltaWt0** [comment]
@MoisesGSalas 

I've updated [appVersion](https://github.com/openedx/openedx-k8s-harmony/pull/31/files#diff-b7bfc0df656c74d6c19d795c70fcc8be3365e95ed993214c9565cdfcc5c7b0a6R14), removed any docker image patching and re-tested everything from scratch - recreated cluster and installed Palm release from a scratch too.

I'm able to
---

1. Create a new course and reindex it

image

2. Create and reindex content library

image


```
[2023-07-01T19:15:23,716][INFO ][o.o.p.PluginsService     ] [harmony-search-cluster-master-2] PluginService:onIndexModule index:[openedx-02-gz4m-library_index/xyBpxOiAQ0GwQqFN1NkGNQ]
[2023-07-01T19:15:23,720][INFO ][o.o.c.m.MetadataMappingService] [harmony-search-cluster-master-2] [openedx-02-gz4m-library_index/xyBpxOiAQ0GwQqFN1NkGNQ] update_mapping [_doc]

```

---

I'm still unable to get `importdemocourse` task to work due to `edx-platform` bug, however it actually imports the course but can't index it:
```
File "/openedx/edx-platform/cms/djangoapps/contentstore/management/commands/reindex_course.py", line 81, in handle
    index_exists = searcher._es.indices.exists(index=index_name)  # pylint: disable=protected-access
    ....
elasticsearch.exceptions.AuthorizationException: AuthorizationException(403, '')
```

---
IMPORTANT: I've rebased the branch on top of latest `main` (project structure has been changed).

**Comment by MoisesGSalas** [comment]
@cmltaWt0, you are 100% right. I didn't even noticed the name changed. I tested again using a Palm image and everything seems to be working.

I think we can go ahead with what we have here.

**Review by MoisesGSalas — APPROVED** [review]

**Comment by cmltaWt0** [comment]
@felipemontoya 

I've re-tested ElasticSearch deployment and pushed one missed setting for the service name (I tested opensearch but forgot to test elastic for any regression).

Retested again for ElasticSearch and for OpenSearch - everything is great.

I'm convinced we can merge it now.

**Comment by felipemontoya** [comment]
Thanks @cmltaWt0, this is wonderful.
I'll merge then.

**Comment by openedx-webhooks** [comment]
@cmltaWt0 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/openedx-k8s-harmony&entry.867055334=https://github.com/openedx/openedx-k8s-harmony/pull/31&entry.1484655318=https://github.com/cmltaWt0&entry.752974735=2023-04-11+14:38&entry.1917517419=2023-07-12+16:15&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by cmltaWt0** [comment]
Thanks @felipemontoya 🚀

**Pull request merged**
